library/std/src/sys/pal/uefi/os.rs

```diff
@@ -124,7 +125,9 @@ pub fn error_string(errno: RawOsError) -> String {
 }
 
 pub fn getcwd() -> io::Result<PathBuf> {
```

Comment:
The UEFI shell protocol actually has a function for this, shouldn't that be tried first?

Reply:
Currently (and even in the original version with more functionality), I was not depending on the availability of the UEFI Shell. The reason was that it's missing from a lot of OEM UEFI.

Also, is there any reason to use a function here? I tried the current implementation, and it works pretty well. And it is always going to be available.

Reply:
Also, after looking a bit into it, I am not sure that the `EFI_SHELL_PROTOCOL.GetCurDir()` returns a text representation that can be used with `EFI_DEVICE_PATH_FROM_TEXT_PROTOCOL` and other non UEFI shell protocols. Feel free to correct me if I am wrong though.

Reply:
👍 Fair enough!

Reply:
@Ayush1325 I definitely agree that you shouldn't count on the shell protocol. However, it does seem unfortunate that running an executable in a context that *does* have a current directory (e.g. from the shell) will always use the dirname of the executable rather than the current directory. That seems like slightly surprising behavior.

Not suggesting that we shouldn't merge this in the interim, but using the shell protocol if available (if we can get the current directory in the form we need) and falling back to this seems like it'd be more usable.

Reply:
> @Ayush1325 I definitely agree that you shouldn't count on the shell protocol. However, it does seem unfortunate that running an executable in a context that _does_ have a current directory (e.g. from the shell) will always use the dirname of the executable rather than the current directory. That seems like slightly surprising behavior.

That's a good point, I did not consider it. Let me remove `getcwd` from this PR and find a better way for it.